cmds/core/gosh/gosh.go

```diff
@@ -58,68 +44,141 @@ func main() {
 	}
 }
 
-func (s shell) runAll(narg int) error {
-	r, err := interp.New(interp.StdIO(os.Stdin, os.Stdout, os.Stderr))
+func run(narg int, args []string) error {
+	runner, err := interp.New(interp.StdIO(os.Stdin, os.Stdout, os.Stderr))
 	if err != nil {
 		return err
 	}
 
 	if narg > 0 {
-		return s.run(r, strings.NewReader(strings.Join(flag.Args(), " ")), "")
+		if strings.HasSuffix(args[0], "sh") {
+			return runScript(runner, args, args[0])
+		}
+
+		return runCmd(runner, strings.NewReader(strings.Join(args, " ")), args[0])
 	}
 
 	if narg == 0 {
 		if term.IsTerminal(int(os.Stdin.Fd())) {
-			return s.runInteractiveTabCompletion(r, os.Stdout)
+			return runInteractive(runner, os.Stdout, os.Stderr)
 		}
 
-		return s.run(r, os.Stdin, "")
+		return runCmd(runner, os.Stdin, "")
 	}
 
 	return nil
 }
 
-func (s shell) run(r *interp.Runner, reader io.Reader, name string) error {
-	prog, err := syntax.NewParser().Parse(reader, name)
+func runScript(runner *interp.Runner, args []string, name string) error {
+	if len(args) > 1 {
+		return fmt.Errorf("no support for trailing arguments to script: %v", args[1:])
+	}
+
+	f, err := os.Open(name)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	prog, err := syntax.NewParser().Parse(f, name)
```

Comment:
FYI you should try to reuse the parser to better reuse allocations. It could be created next to the interp.Runner for example.

Reply:
Makes sense, I'll add a commit here to change that.